src/oss/langchain-messages.mdx

```diff
@@ -2,14 +2,488 @@
+
 ## Tool Message
 
+Tool messages contain the results of tool executions and are used to pass data back to models after external processing.
+
+TODO link out to tool calling conceptual guide
+
+### Basic Tool Messages
+
+```python
+from langchain_core.messages import ToolMessage, AIMessage
+
+# After a model makes a tool call
+ai_message = AIMessage(
+    tool_calls=[{
+        "name": "get_weather",
+        "args": {"location": "San Francisco"},
+        "id": "call_123"
+    }]
+)
+
+# Execute tool and create result message
+weather_result = "Sunny, 72°F"
+tool_message = ToolMessage(
+    content=weather_result,
+    tool_call_id="call_123"  # Must match the call ID
+)
+
+# Continue conversation
+messages = [
+    HumanMessage("What's the weather in San Francisco?"),
+    ai_message,  # Model's tool call
+    tool_message,  # Tool execution result
+]
+response = model.invoke(messages)  # Model processes the result
+```
+
+### Tool Message Structure
+
+```python
+tool_message = ToolMessage(
+    content="Operation completed successfully",
+    tool_call_id="call_123",  # Links to specific tool call
+    name="calculator",  # Optional tool name
+    artifact={"raw_data": [1, 2, 3]}  # Additional data not sent to model
+    # TODO link out to artifact guide ref
+)
+```
+
+### Error Handling in Tool Messages
+
+```python
+# Handle tool execution errors
+try:
+    result = execute_weather_tool(location)
+    tool_message = ToolMessage(
+        content=f"Weather: {result}",
+        tool_call_id=call_id
+    )
+except Exception as e:
+    tool_message = ToolMessage(
+        content=f"Error getting weather: {str(e)}",
+        tool_call_id=call_id
+    )
+```
+
+### Advanced Tool Workflows
+
+```python
+# Multi-step tool execution
+def handle_tool_calls(messages: list, tool_calls: list):
+    """Execute multiple tools and return updated messages"""
+    for tool_call in tool_calls:
+        result = execute_tool(tool_call["name"], tool_call["args"])
+        tool_msg = ToolMessage(
+            content=str(result),
+            tool_call_id=tool_call["id"]
+        )
+        messages.append(tool_msg)
+    return messages
+
+# Use in conversation loop
+messages = [HumanMessage("Calculate 25 * 4 and get weather for NYC")]
+response = llm_with_tools.invoke(messages)
+
+if response.tool_calls:
+    messages.append(response)  # Add AI response with tool calls
+    messages = handle_tool_calls(messages, response.tool_calls)
+    final_response = model.invoke(messages)  # Get final answer
+```
+
 ## Content Blocks
+
+Content blocks enable **provider-agnostic** multimodal interactions. Instead of handling different image formats, audio types, and document structures across providers, content blocks provide a unified interface that works everywhere (supported).
+
+TODO link out to guide
+
+### Why Content Blocks Matter
+
+TODO this needs to be a gif that shows the massive differences at a glance
+
+```python
+# ❌ Provider-specific multimodal handling
```

Comment:
i thought content blocks were more for output? are they also for input?

if yes, should probably have example for both